specs/phase1/beacon-chain.md

```diff
@@ -752,298 +421,108 @@ def process_operations(state: BeaconState, body: BeaconBlockBody) -> None:
-        if data.slot > GENESIS_SLOT:
-            # On-time attestations should have a non-empty shard transition root
-            assert data.shard_transition_root != hash_tree_root(ShardTransition())
-        else:
-            assert data.shard_transition_root == hash_tree_root(ShardTransition())
-    # Type 2: no shard transition
-    else:
-        # Ensure delayed attestation
-        assert data.slot < compute_previous_slot(state.slot)
-        # Late attestations cannot have a shard transition root
-        assert data.shard_transition_root == Root()
-
-    # Signature check
-    assert is_valid_indexed_attestation(state, get_indexed_attestation(state, attestation))
-```
+### New Attestation processing
 
-###### Updated `process_attestation`
+#### Updated `process_attestation`
 
 ```python
 def process_attestation(state: BeaconState, attestation: Attestation) -> None:
-    validate_attestation(state, attestation)
-    # Store pending attestation for epoch processing
-    pending_attestation = PendingAttestation(
-        aggregation_bits=attestation.aggregation_bits,
-        data=attestation.data,
-        inclusion_delay=state.slot - attestation.data.slot,
-        proposer_index=get_beacon_proposer_index(state),
-        crosslink_success=False,  # To be filled in during process_shard_transitions
+    phase0.process_attestation(state, attestation)
+    update_pending_votes(
+        state=state,
+        attestation: Attestation,
+        root=,
+        aggregation_bits=attestation.aggregation_bits
     )
-    if attestation.data.target.epoch == get_current_epoch(state):
-        state.current_epoch_attestations.append(pending_attestation)
-    else:
-        state.previous_epoch_attestations.append(pending_attestation)
 ```
 
-##### Shard transition processing
-
-###### `apply_shard_transition`
+#### `update_pending_votes`
 
 ```python
-def apply_shard_transition(state: BeaconState, shard: Shard, transition: ShardTransition) -> None:
-    # TODO: only need to check it once when phase 1 starts
-    assert state.slot > PHASE_1_FORK_SLOT
-
-    # Correct data root count
-    offset_slots = get_offset_slots(state, shard)
-    assert (
-        len(transition.shard_data_roots)
-        == len(transition.shard_states)
-        == len(transition.shard_block_lengths)
-        == len(offset_slots)
+def update_pending_votes(state: BeaconState,
+                         attestation: Attestation) -> None:
+    if compute_epoch_at_slot(slot) == get_current_epoch(state):
+        pending_headers = state.current_epoch_pending_shard_headers
+    else:
+        pending_headers = state.previous_epoch_pending_shard_headers
+    # Create or update the PendingShardHeader object
+    pending_header = None
+    for header in pending_headers:
+        if header.root == attestation.data.shard_header_root:
+            pending_header = header
+    assert pending_header is not None
+    assert pending_header.slot == attestation.data.slot + 1
```

Comment:
Confused about this check. Why does the attestation have to be on-time? Shouldn't it be - 1 ?

Reply:
```suggestion
    assert pending_header.slot == attestation.data.slot
```

Shouldn't it even be simply `pending_header.slot == attestation.data.slot`?